content/blog/2016/2016-10-11-stage-lock-milestone.adoc

```diff
@@ -0,0 +1,167 @@
+
+[source, groovy]
+----
+stage("Build"){
+  milestone
+  echo "Building"
+}
+stage("Test"){
+  milestone
+  echo "Testing"
+}
+----
+
+When using the `input` step or the `lock` step a backlog of concurrent builds can easily stack up, either waiting for user input or waiting for a resource to become free. The `milestone` step will automatically prune all older jobs that are waiting at these junctions.
+
+[source, groovy]
+----
+milestone
+input message: "Proceed?"
+milestone
+----
+
+Bookending an `input` step like this allows you to select a specific build to proceed and automatically abort all antecedent builds.
+
+[source, groovy]
+----
+milestone
+lock(resource: 'myResource', inversePrecedence: true){
+  echo "locked step"
+  milestone
+}
+----
+
+Similarly a pair of `milestone` steps used with a `lock` will discard all old builds waiting for a shared resource. In this example, `inversePrecedence: true` instructs the `lock` to begin most recent waiting build first, ensuring that the most recent code takes precedence.
+
+=== Putting it all together
+
+Each of these steps can be used independently of the others to control one aspect of a Pipeline or they can be combined to provide powerful, fine-grained control of every aspect of multiple concurrent builds flowing through a Pipeline. Here is a very simple example utilizing all three:
+
+[source, groovy]
+----
+stage("Build"){
+  // The first milestone step starts tracking concurrent build order
+  milestone
+  node{
+    echo "Building"
+  }
+}
+
+// This locked resource contains both Test stages as a single concurrency Unit.
+// Only 1 concurrent build is allowed to utilize the test resources at a time.
+// Newer builds are pulled off the queue first. When a build reaches the
+// milestone at the end of the lock, all jobs started prior to the current
+// build that are still waiting for the lock will be aborted
+lock(resource: 'myResource', inversePrecedence: true){
+  node('test'){
+    stage("Unit Tests"){
+      echo "Unit Tests"
+    }
+    stage("System Tests"){
+      echo "System Tests"
+    }
+  }
+    milestone
+}
+
+// The Deploy stage does not limit concurrency but requires manual input
+// from a user. Several builds might reach this step waiting for input.
+// When a user promotes a specific build all preceding builds are aborted,
+// ensuring that the latest code is always deployed.
+stage("Deploy"){
+  input "Deploy?"
+  milestone
+  node{
+    echo "Deploying"
+  }
+}
+----
+
+For a more complete and complex example utilizing all these steps in a Pipeline check out the  link:https://github.com/jenkinsci/workflow-aggregator-plugin/blob/master/demo/repo/Jenkinsfile[Jenkinsfile] provided with the link:https://github.com/jenkinsci/workflow-aggregator-plugin/tree/master/demo[Docker image for demonstrating Pipeline]. This is a working demo that can be quickly set up and run.
```

Comment:
Would recommend using permalinks (`Y` key in GitHub) to ensure that the URLs do not break after internal refactorings.